File: tui/content.py
"""The content corpus: frontmatter parsing and the generated index.

``content/`` is the single source of truth for the whole tutor. Two very
different readers consume it:

* this TUI, which needs an ordered navigation tree, and
* the Claude Code skill, which needs a lookup table to find the one file that
  answers a question.

Both read ``content/index.json``, generated here. Neither copies prose, so
there is exactly one place a fact can live and exactly one place to edit it.

The index rebuilds itself whenever any article is newer than it, so authoring
is "drop a markdown file in and reopen" — there is no build step to forget.
"""

import json
import os
import re

INDEX_NAME = "index.json"

# Frontmatter is a deliberately tiny subset: `key: scalar` and `key: [a, b]`.
# Hand-parsing it avoids a PyYAML dependency, which would otherwise be the
# only thing standing between her and a working install.
_FM_RE = re.compile(r"\A---\r?\n(.*?)\r?\n---\r?\n?", re.S)
_KEY_RE = re.compile(r"^([A-Za-z_][A-Za-z0-9_-]*)\s*:\s*(.*)$")
_INT_KEYS = ("order",)
_LIST_KEYS = ("keywords",)


def _unquote(value):
    if len(value) >= 2 and value[0] == value[-1] and value[0] in "\"'":
        return value[1:-1]
    return value


def parse_frontmatter(text):
    """Return ``(meta, body)``. Missing or malformed frontmatter yields ``{}``."""
    m = _FM_RE.match(text)
    if not m:
        return {}, text
    meta = {}
    for line in m.group(1).split("\n"):
        line = line.rstrip()
        if not line.strip() or line.lstrip().startswith("#"):
            continue
        km = _KEY_RE.match(line)
        if not km:
            continue
        key, raw = km.group(1), km.group(2).strip()
        if raw.startswith("[") and raw.endswith("]"):
            inner = raw[1:-1].strip()
            meta[key] = [_unquote(v.strip()) for v in inner.split(",") if v.strip()]
        else:
            value = _unquote(raw)
            if key in _INT_KEYS:
                try:
                    value = int(value)
                except ValueError:
                    pass
            meta[key] = value
    for key in _LIST_KEYS:
        if isinstance(meta.get(key), str):
            meta[key] = [v.strip() for v in meta[key].split(",") if v.strip()]
    return meta, text[m.end():]


def deslug(slug):
    """``01-the-terminal`` -> ``The Terminal``. A fallback, not the rule."""
    slug = re.sub(r"^\d+[-_]", "", slug)
    words = re.split(r"[-_\s]+", slug)
    return " ".join(w[:1].upper() + w[1:] for w in words if w)


def _sort_key(prefix_source, order, title):
    """Order by explicit `order`, then by filename prefix, then by title."""
    m = re.match(r"^(\d+)", prefix_source)
    numeric = int(m.group(1)) if m else 10 ** 6
    return (order if isinstance(order, int) else numeric, numeric, title.lower())


def content_dir(root):
    return os.path.join(root, "content")


def index_path(root):
    return os.path.join(content_dir(root), INDEX_NAME)


def iter_markdown(root):
    base = content_dir(root)
    if not os.path.isdir(base):
        return
    for slug in sorted(os.listdir(base)):
        part_dir = os.path.join(base, slug)
        if not os.path.isdir(part_dir) or slug.startswith((".", "_")):
            continue
        for name in sorted(os.listdir(part_dir)):
            if name.endswith(".md") and not name.startswith((".", "_")):
                yield slug, name, os.path.join(part_dir, name)


def build_index(root):
    """Scan ``content/`` and return the index structure."""
    parts = {}
    for slug, name, path in iter_markdown(root):
        try:
            with open(path, encoding="utf-8") as fh:
                text = fh.read()
        except OSError:
            continue
        meta, body = parse_frontmatter(text)

        title = meta.get("title") or _first_heading(body) or deslug(name[:-3])
        part_title = meta.get("part") or deslug(slug)
        article = {
            "id": meta.get("id") or "%s/%s" % (re.sub(r"^\d+[-_]", "", slug), name[:-3]),
            "title": title,
            "section": meta.get("section", ""),
            "summary": meta.get("summary", ""),
            "keywords": meta.get("keywords", []),
            "path": os.path.relpath(path, root).replace(os.sep, "/"),
            "_sort": _sort_key(name, meta.get("order"), title),
        }
        part = parts.setdefault(
            slug,
            {"slug": slug, "title": part_title, "level": meta.get("level", ""), "articles": []},
        )
        # The first article carrying an explicit `part:` names the part; later
        # ones do not get to rename it out from under the earlier files.
        if meta.get("part") and part["title"] == deslug(slug):
            part["title"] = meta["part"]
        # `level:` follows the same first-one-wins rule.
        if not part["level"]:
            part["level"] = meta.get("level", "")
        part["articles"].append(article)

    ordered = []
    for slug in sorted(parts, key=lambda s: _sort_key(s, None, parts[s]["title"])):
        part = parts[slug]
        part["articles"].sort(key=lambda a: a["_sort"])
        for article in part["articles"]:
            del article["_sort"]
        ordered.append(part)
    return {"parts": ordered}


def _first_heading(body):
    for line in body.split("\n"):
        m = re.match(r"^#{1,6}\s+(.*?)\s*#*\s*$", line)
        if m:
            return m.group(1).strip()
    return None


def write_index(root):
    index = build_index(root)
    target = index_path(root)
    os.makedirs(os.path.dirname(target), exist_ok=True)
    with open(target, "w", encoding="utf-8") as fh:
        json.dump(index, fh, indent=2, ensure_ascii=False)
        fh.write("\n")
    return index


def is_stale(root):
    """True when any article, or the set of articles, is newer than the index."""
    target = index_path(root)
    if not os.path.exists(target):
        return True
    stamp = os.path.getmtime(target)
    base = content_dir(root)
    if os.path.isdir(base) and os.path.getmtime(base) > stamp:
        return True
    for _slug, _name, path in iter_markdown(root):
        if os.path.getmtime(path) > stamp:
            return True
        # A rename or deletion changes the directory's own mtime.
        if os.path.getmtime(os.path.dirname(path)) > stamp:
            return True
    return False


def load_index(root, rebuild=True):
    """Load the index, regenerating it first if the content has moved on."""
    if rebuild and is_stale(root):
        try:
            return write_index(root)
        except OSError:
            # A read-only install still works; it just cannot cache.
            return build_index(root)
    try:
        with open(index_path(root), encoding="utf-8") as fh:
            return json.load(fh)
    except (OSError, ValueError):
        return build_index(root)


def read_article(root, article):
    """Return the markdown body of *article* with frontmatter stripped."""
    path = os.path.join(root, article["path"])
    try:
        with open(path, encoding="utf-8") as fh:
            text = fh.read()
    except OSError as exc:
        return "# Missing article\n\n`%s` could not be read.\n\n    %s\n" % (
            article.get("path", "?"),
            exc,
        )
    _meta, body = parse_frontmatter(text)
    return body


def part_sections(part):
    """Group a part's articles into ``(title, from, to)`` runs.

    A section is a run of consecutive articles sharing a ``section:`` value,
    the empty one included. It is derived here rather than stored, so the
    index stays a flat list of articles and search, :func:`flatten` and the
    Claude Code skill all carry on reading it exactly as before. A part where
    no article carries ``section:`` is a single untitled run spanning the
    lot, which draws as the plain numbered list it has always been.
    """
    out = []
    for i, article in enumerate(part.get("articles", [])):
        title = article.get("section", "")
        if out and out[-1][0] == title and out[-1][2] == i:
            out[-1] = (title, out[-1][1], i + 1)
            continue
        out.append((title, i, i + 1))
    return out


def section_at(part, i):
    """The index of the section holding article *i*, or ``None``."""
    for si, (_title, start, stop) in enumerate(part_sections(part)):
        if start <= i < stop:
            return si
    return None


def level_of(part):
    """A part's level, falling back to its own title.

    That fallback is what makes a corpus carrying no ``level:`` at all
    degrade to the old behaviour exactly — every part becomes its own
    single-part level, so the tab bar still reads as the list of parts and
    the sidebar suppresses the redundant part heading.
    """
    return part.get("level") or part["title"]


def index_levels(index):
    """Group the index's parts into runs of equal ``level:`` as ``(title, from, to)``.

    A level is a run of consecutive parts sharing a ``level:`` value — the
    tabs along the top. It is derived the same way a section is, and for the
    same reason: the index stays a flat list of parts holding a flat list of
    articles, so search, :func:`flatten` and the Claude Code skill never
    learnt that levels exist.
    """
    out = []
    for i, part in enumerate(index.get("parts", [])):
        title = level_of(part)
        if out and out[-1][0] == title and out[-1][2] == i:
            out[-1] = (title, out[-1][1], i + 1)
            continue
        out.append((title, i, i + 1))
    return out


def level_at(index, i):
    """The index of the level holding part *i*, or ``None``."""
    for li, (_title, start, stop) in enumerate(index_levels(index)):
        if start <= i < stop:
            return li
    return None


def flatten(index):
    """All articles in reading order as ``(part_i, article_i, part, article)``."""
    out = []
    for pi, part in enumerate(index.get("parts", [])):
        for ai, article in enumerate(part.get("articles", [])):
            out.append((pi, ai, part, article))
    return out


def search(root, index, query):
    """Rank articles against *query*. Title and keyword hits outrank body hits."""
    query = query.strip().lower()
    if not query:
        return []
    terms = query.split()
    results = []
    for pi, ai, part, article in flatten(index):
        haystacks = (
            article["title"].lower(),
            " ".join(article.get("keywords", [])).lower(),
            article.get("summary", "").lower(),
            read_article(root, article).lower(),
        )
        weights = (10, 6, 4, 1)
        score = 0
        for term in terms:
            hit = 0
            for hay, weight in zip(haystacks, weights):
                if term in hay:
                    hit = max(hit, weight)
            if not hit:
                score = 0
                break  # every term must appear somewhere
            score += hit
        if score:
            results.append((score, pi, ai, part, article))
    results.sort(key=lambda r: (-r[0], r[1], r[2]))
    return [(pi, ai, part, article) for _s, pi, ai, part, article in results]

File: go/content.go
// The content corpus: frontmatter parsing and the generated index.
//
// content/ is the single source of truth for the whole tutor. Two very
// different readers consume it:
//
//   - this TUI, which needs an ordered navigation tree, and
//   - the Claude Code skill, which needs a lookup table to find the one file
//     that answers a question.
//
// Both read content/index.json, generated here. Neither copies prose, so
// there is exactly one place a fact can live and exactly one place to edit it.
//
// The index rebuilds itself whenever any article is newer than it, so
// authoring is "drop a markdown file in and reopen" — there is no build step
// to forget.
package main

import (
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

const indexName = "index.json"

// Article, Part and Index are laid out so encoding/json emits fields in the
// same order Python's dict insertion does. The parity check is byte for byte,
// so field order is load-bearing.
type Article struct {
	ID       string   `json:"id"`
	Title    string   `json:"title"`
	Section  string   `json:"section"`
	Summary  string   `json:"summary"`
	Version  string   `json:"version"`
	Keywords []string `json:"keywords"`
	Path     string   `json:"path"`

	sort sortKey
}

type Part struct {
	Slug     string    `json:"slug"`
	Title    string    `json:"title"`
	Level    string    `json:"level"`
	Articles []Article `json:"articles"`
}

// Section is a run of consecutive articles sharing a `section:` value. It is
// derived at render time rather than stored: the index stays a flat list of
// articles, so search, `flatten` and the Claude Code skill all carry on
// reading it exactly as before.
type Section struct {
	Title    string
	From, To int // [From, To)
}

func (s Section) contains(i int) bool { return s.From <= i && i < s.To }

// partSections groups a part's articles into runs of equal `section:`,
// including the empty one. A part where no article carries `section:` is
// therefore a single untitled run spanning the lot, which draws as the plain
// numbered list it has always been — the sectionless case is not a special
// case, it is the degenerate one.
func partSections(part Part) []Section {
	out := []Section{}
	for i, a := range part.Articles {
		if n := len(out); n > 0 && out[n-1].Title == a.Section && out[n-1].To == i {
			out[n-1].To = i + 1
			continue
		}
		out = append(out, Section{a.Section, i, i + 1})
	}
	return out
}

// sectionAt returns the section holding article i, and whether there is one.
func sectionAt(part Part, i int) (int, bool) {
	for si, s := range partSections(part) {
		if s.contains(i) {
			return si, true
		}
	}
	return 0, false
}

type Index struct {
	Parts []Part `json:"parts"`
}

// Level is a run of consecutive parts sharing a `level:` value — the tabs
// along the top. It is derived the same way Section is, and for the same
// reason: the index stays a flat list of parts holding a flat list of
// articles, so search, `flatten` and the Claude Code skill never learnt that
// levels exist.
type Level struct {
	Title    string
	From, To int // [From, To) over index.Parts
}

func (l Level) contains(i int) bool { return l.From <= i && i < l.To }

// levelOf is a part's level, falling back to its own title. That fallback is
// what makes a corpus carrying no `level:` at all degrade to the old
// behaviour exactly — every part becomes its own single-part level, so the
// tab bar still reads as the list of parts and the sidebar suppresses the
// redundant part heading.
func levelOf(part Part) string {
	if part.Level != "" {
		return part.Level
	}
	return part.Title
}

// indexLevels groups the index's parts into runs of equal `level:`.
func indexLevels(index Index) []Level {
	out := []Level{}
	for i, p := range index.Parts {
		if n := len(out); n > 0 && out[n-1].Title == levelOf(p) && out[n-1].To == i {
			out[n-1].To = i + 1
			continue
		}
		out = append(out, Level{levelOf(p), i, i + 1})
	}
	return out
}

// levelAt returns the level holding part i, and whether there is one.
func levelAt(index Index, i int) (int, bool) {
	for li, l := range indexLevels(index) {
		if l.contains(i) {
			return li, true
		}
	}
	return 0, false
}

type sortKey struct {
	primary int
	numeric int
	title   string
}

func (a sortKey) less(b sortKey) bool {
	if a.primary != b.primary {
		return a.primary < b.primary
	}
	if a.numeric != b.numeric {
		return a.numeric < b.numeric
	}
	return a.title < b.title
}

// Frontmatter is a deliberately tiny subset: `key: scalar` and `key: [a, b]`.
// Hand-parsing it avoids a YAML dependency, which would otherwise be the only
// thing standing between her and a working install.
var (
	fmRe      = regexp.MustCompile(`(?s)\A---\r?\n(.*?)\r?\n---\r?\n?`)
	keyRe     = regexp.MustCompile(`^([A-Za-z_][A-Za-z0-9_-]*)\s*:\s*(.*)$`)
	numRe     = regexp.MustCompile(`^(\d+)`)
	prefixRe  = regexp.MustCompile(`^\d+[-_]`)
	headRe    = regexp.MustCompile(`^#{1,6}\s+(.*?)\s*#*\s*$`)
	versionRe = regexp.MustCompile(`(?m)^\*(v[0-9]+(?:\.[0-9]+)*)\*$`)
)

type frontmatter struct {
	vals  map[string]string
	lists map[string][]string
	order *int
}

func (f frontmatter) get(key string) string { return f.vals[key] }

func (f frontmatter) keywords() []string {
	if l, ok := f.lists["keywords"]; ok {
		return l
	}
	if s, ok := f.vals["keywords"]; ok {
		out := []string{}
		for _, v := range strings.Split(s, ",") {
			if t := strings.TrimSpace(v); t != "" {
				out = append(out, t)
			}
		}
		return out
	}
	return []string{}
}

func unquote(value string) string {
	r := []rune(value)
	if len(r) >= 2 && r[0] == r[len(r)-1] && (r[0] == '"' || r[0] == '\'') {
		return string(r[1 : len(r)-1])
	}
	return value
}

// parseFrontmatter returns the metadata and the body. Missing or malformed
// frontmatter yields empty metadata and the text untouched.
func parseFrontmatter(text string) (frontmatter, string) {
	meta := frontmatter{vals: map[string]string{}, lists: map[string][]string{}}
	loc := fmRe.FindStringSubmatchIndex(text)
	if loc == nil || loc[0] != 0 {
		return meta, text
	}
	block := text[loc[2]:loc[3]]
	for _, line := range strings.Split(block, "\n") {
		line = strings.TrimRightFunc(line, isSpaceRune)
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}
		km := keyRe.FindStringSubmatch(line)
		if km == nil {
			continue
		}
		key, raw := km[1], strings.TrimSpace(km[2])
		if strings.HasPrefix(raw, "[") && strings.HasSuffix(raw, "]") {
			inner := strings.TrimSpace(raw[1 : len(raw)-1])
			list := []string{}
			for _, v := range strings.Split(inner, ",") {
				if strings.TrimSpace(v) != "" {
					list = append(list, unquote(strings.TrimSpace(v)))
				}
			}
			meta.lists[key] = list
			delete(meta.vals, key)
			continue
		}
		value := unquote(raw)
		if key == "order" {
			if n, err := strconv.Atoi(value); err == nil {
				meta.order = &n
			}
		}
		meta.vals[key] = value
		delete(meta.lists, key)
	}
	return meta, text[loc[1]:]
}

// deslug turns `01-the-terminal` into `The Terminal`. A fallback, not the rule.
func deslug(slug string) string {
	slug = prefixRe.ReplaceAllString(slug, "")
	words := regexp.MustCompile(`[-_\s]+`).Split(slug, -1)
	out := []string{}
	for _, w := range words {
		if w == "" {
			continue
		}
		r := []rune(w)
		out = append(out, strings.ToUpper(string(r[0]))+string(r[1:]))
	}
	return strings.Join(out, " ")
}

// makeSortKey orders by explicit `order`, then by filename prefix, then title.
func makeSortKey(prefixSource string, order *int, title string) sortKey {
	numeric := 1000000
	if m := numRe.FindStringSubmatch(prefixSource); m != nil {
		if n, err := strconv.Atoi(m[1]); err == nil {
			numeric = n
		}
	}
	primary := numeric
	if order != nil {
		primary = *order
	}
	return sortKey{primary, numeric, strings.ToLower(title)}
}

func contentDir(root string) string { return filepath.Join(root, "content") }
func indexPath(root string) string  { return filepath.Join(contentDir(root), indexName) }

type mdFile struct {
	slug, name, path string
}

func iterMarkdown(root string) []mdFile {
	out := []mdFile{}
	base := contentDir(root)
	entries, err := os.ReadDir(base)
	if err != nil {
		return out
	}
	slugs := []string{}
	for _, e := range entries {
		slugs = append(slugs, e.Name())
	}
	sort.Strings(slugs)
	for _, slug := range slugs {
		partDir := filepath.Join(base, slug)
		info, err := os.Stat(partDir)
		if err != nil || !info.IsDir() || strings.HasPrefix(slug, ".") || strings.HasPrefix(slug, "_") {
			continue
		}
		files, err := os.ReadDir(partDir)
		if err != nil {
			continue
		}
		names := []string{}
		for _, f := range files {
			names = append(names, f.Name())
		}
		sort.Strings(names)
		for _, name := range names {
			if strings.HasSuffix(name, ".md") &&
				!strings.HasPrefix(name, ".") && !strings.HasPrefix(name, "_") {
				out = append(out, mdFile{slug, name, filepath.Join(partDir, name)})
			}
		}
	}
	return out
}

func firstHeading(body string) string {
	for _, line := range strings.Split(body, "\n") {
		if m := headRe.FindStringSubmatch(line); m != nil {
			return strings.TrimSpace(m[1])
		}
	}
	return ""
}

// extractVersion returns an article's own version tag — the italic line
// between the H1 and the first paragraph, e.g. `*v0.2.0*` — including its
// leading "v", or the empty string if no line in the body matches. Only the
// first matching line counts; render.go passes the line through as ordinary
// italic text and must go on doing so, so this never touches rendering.
func extractVersion(body string) string {
	if m := versionRe.FindStringSubmatch(body); m != nil {
		return m[1]
	}
	return ""
}

// buildIndex scans content/ and returns the index structure.
func buildIndex(root string) Index {
	type partAcc struct {
		part *Part
		key  sortKey
	}
	order := []string{}
	parts := map[string]*Part{}

	for _, f := range iterMarkdown(root) {
		raw, err := os.ReadFile(f.path)
		if err != nil {
			continue
		}
		meta, body := parseFrontmatter(string(raw))

		stem := f.name[:len(f.name)-3]
		title := meta.get("title")
		if title == "" {
			title = firstHeading(body)
		}
		if title == "" {
			title = deslug(stem)
		}
		partTitle := meta.get("part")
		if partTitle == "" {
			partTitle = deslug(f.slug)
		}
		id := meta.get("id")
		if id == "" {
			id = prefixRe.ReplaceAllString(f.slug, "") + "/" + stem
		}
		rel, err := filepath.Rel(root, f.path)
		if err != nil {
			rel = f.path
		}
		article := Article{
			ID:       id,
			Title:    title,
			Section:  meta.get("section"),
			Summary:  meta.get("summary"),
			Version:  extractVersion(body),
			Keywords: meta.keywords(),
			Path:     filepath.ToSlash(rel),
			sort:     makeSortKey(f.name, meta.order, title),
		}

		part, ok := parts[f.slug]
		if !ok {
			part = &Part{Slug: f.slug, Title: partTitle, Level: meta.get("level"),
				Articles: []Article{}}
			parts[f.slug] = part
			order = append(order, f.slug)
		}
		// The first article carrying an explicit `part:` names the part; later
		// ones do not get to rename it out from under the earlier files.
		if meta.get("part") != "" && part.Title == deslug(f.slug) {
			part.Title = meta.get("part")
		}
		// `level:` follows the same first-one-wins rule.
		if part.Level == "" {
			part.Level = meta.get("level")
		}
		part.Articles = append(part.Articles, article)
	}

	accs := []partAcc{}
	for _, slug := range order {
		accs = append(accs, partAcc{parts[slug], makeSortKey(slug, nil, parts[slug].Title)})
	}
	sort.SliceStable(accs, func(i, j int) bool { return accs[i].key.less(accs[j].key) })

	index := Index{Parts: []Part{}}
	for _, acc := range accs {
		arts := acc.part.Articles
		sort.SliceStable(arts, func(i, j int) bool { return arts[i].sort.less(arts[j].sort) })
		acc.part.Articles = arts
		index.Parts = append(index.Parts, *acc.part)
	}
	return index
}

// encodeIndex matches Python's json.dump(indent=2, ensure_ascii=False)
// exactly: two-space indent, no HTML escaping, one trailing newline.
func encodeIndex(index Index) ([]byte, error) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	enc.SetIndent("", "  ")
	if err := enc.Encode(index); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

func writeIndex(root string) (Index, error) {
	index := buildIndex(root)
	data, err := encodeIndex(index)
	if err != nil {
		return index, err
	}
	target := indexPath(root)
	if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
		return index, err
	}
	return index, os.WriteFile(target, data, 0o644)
}

// isStale is true when any article, or the set of articles, is newer than the
// index.
func isStale(root string) bool {
	target := indexPath(root)
	st, err := os.Stat(target)
	if err != nil {
		return true
	}
	stamp := st.ModTime()
	if base, err := os.Stat(contentDir(root)); err == nil && base.ModTime().After(stamp) {
		return true
	}
	for _, f := range iterMarkdown(root) {
		if fi, err := os.Stat(f.path); err == nil && fi.ModTime().After(stamp) {
			return true
		}
		// A rename or deletion changes the directory's own mtime.
		if di, err := os.Stat(filepath.Dir(f.path)); err == nil && di.ModTime().After(stamp) {
			return true
		}
	}
	return false
}

// loadIndex loads the index, regenerating it first if the content has moved on.
func loadIndex(root string, rebuild bool) Index {
	if rebuild && isStale(root) {
		index, err := writeIndex(root)
		if err != nil {
			// A read-only install still works; it just cannot cache.
			return buildIndex(root)
		}
		return index
	}
	data, err := os.ReadFile(indexPath(root))
	if err != nil {
		return buildIndex(root)
	}
	var index Index
	if err := json.Unmarshal(data, &index); err != nil {
		return buildIndex(root)
	}
	return index
}

// readArticle returns the markdown body of an article, frontmatter stripped.
func readArticle(root string, a Article) string {
	path := filepath.Join(root, a.Path)
	raw, err := os.ReadFile(path)
	if err != nil {
		return "# Missing article\n\n`" + a.Path + "` could not be read.\n\n    " +
			err.Error() + "\n"
	}
	_, body := parseFrontmatter(string(raw))
	return body
}

type located struct {
	partI, articleI int
	part            Part
	article         Article
}

// flatten lists all articles in reading order.
func flatten(index Index) []located {
	out := []located{}
	for pi, part := range index.Parts {
		for ai, article := range part.Articles {
			out = append(out, located{pi, ai, part, article})
		}
	}
	return out
}

// searchIndex ranks articles against a query. Title and keyword hits outrank
// body hits, and every term must appear somewhere.
func searchIndex(root string, index Index, query string) []located {
	query = strings.ToLower(strings.TrimSpace(query))
	if query == "" {
		return []located{}
	}
	terms := strings.Fields(query)

	type scored struct {
		score int
		item  located
	}
	results := []scored{}

	for _, item := range flatten(index) {
		haystacks := []string{
			strings.ToLower(item.article.Title),
			strings.ToLower(strings.Join(item.article.Keywords, " ")),
			strings.ToLower(item.article.Summary),
			strings.ToLower(readArticle(root, item.article)),
		}
		weights := []int{10, 6, 4, 1}
		score := 0
		for _, term := range terms {
			hit := 0
			for i, hay := range haystacks {
				if strings.Contains(hay, term) && weights[i] > hit {
					hit = weights[i]
				}
			}
			if hit == 0 {
				score = 0
				break
			}
			score += hit
		}
		if score > 0 {
			results = append(results, scored{score, item})
		}
	}

	sort.SliceStable(results, func(i, j int) bool {
		a, b := results[i], results[j]
		if a.score != b.score {
			return a.score > b.score
		}
		if a.item.partI != b.item.partI {
			return a.item.partI < b.item.partI
		}
		return a.item.articleI < b.item.articleI
	})

	out := []located{}
	for _, r := range results {
		out = append(out, r.item)
	}
	return out
}
